### ISSUE TITLE
[FEATURE REQUEST] Add ability to provide a base `context` for `handler` invocations

### ISSUE THREAD

**Issue by binaek** [hibiken/asynq#381]
**Is your feature request related to a problem? Please describe.**
I am unable to set a base `context` for the `handlerFunc` invocations.

**Describe the solution you'd like**
Something along the lines of what `http.Server` provides:
```
BaseContext func(net.Listener) context.Context
```

A nice adaptation would be `BaseContext func(context.Context) context.Context`.

This will allow invocation contexts to be derived from a known `context.Context`

**Comment by hibiken** [comment]
@binaek Thank you for creating this feature request!

I think it makes sense to add this `BaseContext` to `Config`. 
But before we make the change, could you describe possible use cases?  So that we have more concrete motivation to make this change :)

**Comment by binaek** [comment]
@hibiken 
The use case that I am struggling with is as such:

The application uses Context as a DI container for application wide services like DB connection, Root loggers, telemetry etc.

The context also uses a signal.NotifyContext   and so deadlines and cancellations never reach the handlers till an explicit call to Stop or Shutdown

**Comment by hibiken** [comment]
I see. I'm onboard with adding this field to `Config`.

Just to clarify, the proposed function's signature is:

```go
BaseContext func(ctx context.Context) context.Context
```

The input to the `BaseContext` is the context that would normally be passed to Handler (in the absence of this `BaseContext` function provided by the user), is that correct? So in effect this `BaseContext` function works like a middleware where user gets an opportunity to decorate the context before its passed to the Handler invocation? Let me know if I misunderstood your proposal.

**Comment by hibiken** [comment]
Closed via #386
